### ISSUE TITLE
OOM Error after 1 epoch or in the middle of the epoch

### ISSUE THREAD

**Issue by working12** [HuguesTHOMAS/KPConv-PyTorch#181]
Usually, when using this code, I was getting calibration issues and things like that. But now, after the calibration passes correctly, I notice an "OOM error" in the middle of the epoch. 
```

e001-i0135 => L=3.100 acc= 69% / t(ms):  15.1 1081.0 657.5)
error: Detected 1 oom-kill event(s) in StepId=889106.0. Some of your processes may have been killed by the cgroup out-of-memory handler.
e001-i0136 => L=2.946 acc= 72% / t(ms):  15.0 1096.7 661.1)srun: error: node11: task 0: Out Of Memory
error: Detected 1 oom-kill event(s) in StepId=889106.batch. Some of your processes may have been killed by the cgroup out-of-memory handler.
```
1. Usually the amount of memory (GPU memory and/or RAM) needed by code in each epoch is fixed. So if epoch 1 passed correctly it should not affect the next epochs as well. Is there any reason for this?

2. Even with `batch_num = 4`, and `dl=0.02/0.03` I got this error (had close to 32G memory in both GPU and RAM). I have `epoch_steps=500`  do you think reducing this might be helpful? (since as per #180 this controls the number of spheres in total seen by the model in each iteration)?

What are the steps I should follow to solve this kind of issue?

**Comment by HuguesTHOMAS** [comment]
First a question. Is it a frame dataset like SemanticKitti or an area like S3DIS. Can you post a picture of the data?

Here are some suggestions:

 - You can use the lowest batch_num you can (3 is a good guess).
 - You can try to reduce the number of neighbor points in a convolution by reducing `conv_radius`. This will surely mean a loss in performance though.
 - Try to get the highest `dl` without losing too much performance. Try 0.03, maybe 0.032, 0.034...

Everything is a question of trade-off between speed and performance.

Also, consider there might be a bug in your code that makes the memory grow. This is always a possibility if you modified some stuff because of your new data. Does the memory on your GPU increase constantly during training or is it random? 

Another possibility: Maybe you have a part of your dataset that is particularly more dense than the other and that makes everything crash?

**Comment by working12** [comment]
Memory issues in the middle of the epoch were due to one of my changes. 

Regarding the performance, both of the things mentioned would be bad (increasing `dl` and increasing `conv_radius`). Also, dataset density is non-uniform. (Lots of points are in the front - and few points are far from the camera - so on a side view it would look like this)

**Comment by liuxuzbg** [comment]
> Memory issues in the middle of the epoch were due to one of my changes.
> 
> Regarding the performance, both of the things mentioned would be bad (increasing `dl` and increasing `conv_radius`). Also, dataset density is non-uniform. (Lots of points are in the front - and few points are far from the camera - so on a side view it would look like this) @HuguesTHOMAS
I also meet the OOM error, Does the code support multi gpu trainning?
